File: 0013-roman-to-integer/0013-roman-to-integer.java
import java.util.HashMap;

class Solution {
    
    public int romanToInt(String s) {
        HashMap<Character, Integer> romanNumerals = new HashMap<Character, Integer>();
        romanNumerals.put('I', 1);
        romanNumerals.put('V', 5);
        romanNumerals.put('X', 10);
        romanNumerals.put('L', 50);
        romanNumerals.put('C', 100);
        romanNumerals.put('D', 500);
        romanNumerals.put('M', 1000);
        
        char[] charsRev = new StringBuilder(s).reverse().toString().toCharArray();
        System.out.println(charsRev);
        
        int sum = 0;
        int i = 0;
        while (i < charsRev.length) {
            if (i < charsRev.length - 1 && romanNumerals.get(charsRev[i]) > romanNumerals.get(charsRev[i+1])) {
                sum += romanNumerals.get(charsRev[i]) - romanNumerals.get(charsRev[i+1]);
                i += 2;
            } else {
                sum += romanNumerals.get(charsRev[i]);
                i++;
            }
        }
        
        return sum;
    }
}

File: 0013-roman-to-integer/0013-roman-to-integer.py
def map_char(c: str) -> int:
    match c:
        case 'I':
            return 1
        case 'V':
            return 5
        case 'X':
            return 10
        case 'L':
            return 50
        case 'C':
            return 100
        case 'D':
            return 500
        case 'M':
            return 1000
        case _:
            return

class Solution:
    def romanToInt(self, s: str) -> int:
        sum = 0
        i = 0
        while i < len(s):
            curr = s[i]
            if (i < len(s) - 1):
                nxt = s[i+1]
                if (map_char(curr) < map_char(nxt)):
                    sum += map_char(nxt) - map_char(curr)
                    i += 2
                else:
                    sum += map_char(curr)
                    i += 1
            else:
                sum += map_char(curr)
                i += 1
        return sum
        